File: src/tutorial/go2_tutorial/src/go2_cruising_service.cpp
/*
需求：处理客户端提交的数据（0或非0），
    如果是0，停止巡航，如果非，开始巡航
    不过提交的什么数据都需要响应机器人的位置信息。
流程：
    1.创建服务端；
    2.回调函数处理请求，分情况处理（控制机器人巡航），最后响应结果。   
*/

#include <rclcpp/rclcpp.hpp>
#include <go2_tutorial_inter/srv/cruising.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <sport_model.hpp>


using namespace std::chrono_literals;

class Go2CrusingService: public rclcpp::Node{
public:
    Go2CrusingService(const std::string &node_name):Node(node_name){
         RCLCPP_INFO(this->get_logger(),"Go2CruisingService 创建");

        // 创建服务端
        cru_service_ = this->create_service<go2_tutorial_inter::srv::Cruising>("cruising", 
            std::bind(&Go2CrusingService::cru_cb, this, std::placeholders::_1, std::placeholders::_2));

        // 通过订阅里程计获取机器人的位置信息
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>("/odom", 10, 
            std::bind(&Go2CrusingService::odom_cb, this, std::placeholders::_1));

        // 创建远程参数客户端
        param_client_ = std::make_shared<rclcpp::AsyncParametersClient>(this, "go2_ctrl");
        // 让客户端连接服务端
        while (!param_client_->wait_for_service(1s)) {
            if(!rclcpp::ok()){
                RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
                return;
            }
            RCLCPP_INFO(this->get_logger(), "参数服务端go2_ctrl连接中...");
        }
        RCLCPP_INFO(this->get_logger(), "成功连接到参数服务端go2_ctrl!");

        // 参数
        this->declare_parameter("x", 0.1);
        this->declare_parameter("y", 0.0);
        this->declare_parameter("z", 0.5);
    }
private:
    // void cru_cb(
    //     const std::shared_ptr<go2_tutorial_inter::srv::Cruising::Request> request,
    //     const std::shared_ptr<go2_tutorial_inter::srv::Cruising::Response> response){    
    // }
    void cru_cb(const go2_tutorial_inter::srv::Cruising::Request::SharedPtr request,
                const go2_tutorial_inter::srv::Cruising::Response::SharedPtr response){
        // 处理请求
        auto flag = request->flag;
        // 向go2_ctrl节点设置参数, 如果是0, api设置为stopmove, 否则设置为sport
        int32_t id;
        if(flag != 0){
            id = ROBOT_SPORT_API_ID_MOVE;
            RCLCPP_INFO(this->get_logger(),"开始巡航...");
            
        }else{
            id = ROBOT_SPORT_API_ID_STOPMOVE;
            RCLCPP_INFO(this->get_logger(),"停止巡航...");
            
        }
        // 向巡航节点设置参数
        param_client_->set_parameters({
            // rclcpp::Parameter("x", 0.1),
            // rclcpp::Parameter("y", 0.0),
            // rclcpp::Parameter("z", 0.5),
            this->get_parameter("x"),   // 本节点的参数
            this->get_parameter("y"),
            this->get_parameter("z"),
            rclcpp::Parameter("api", id) // go2_ctrl节点的sport_api_id参数
        });

        // 生成响应
        response->point = current_point_;
    }

    void odom_cb(const nav_msgs::msg::Odometry::SharedPtr odom){
        // 获取机器人的位置信息
        current_point_ = odom->pose.pose.position;
    }

    rclcpp::Service<go2_tutorial_inter::srv::Cruising>::SharedPtr cru_service_;
    geometry_msgs::msg::Point current_point_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    // 参数客户端
     rclcpp::AsyncParametersClient::SharedPtr param_client_;
};
 
int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Go2CrusingService>("go2_crusing_service");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/tutorial/go2_tutorial_py/go2_tutorial_py/go2_cruising_service.py
"""
需求： 处理客户端提交的数据(0 或 非0)
    - 0  ：停止巡航
    - 非0: 开始巡航
    无论提交什么数据，都需要响应机器人的位置信息
"""

import rclpy
from rclpy.node import Node
from go2_tutorial_inter.srv import Cruising
from geometry_msgs.msg import Point
from nav_msgs.msg import Odometry
from unitree_api.msg import Request
from .sport_model import ROBOT_SPORT_API_IDS
import json


class Go2CruisingServicePy(Node):

    def __init__(self, node_name: str):
        super().__init__(node_name)
        self.get_logger().info('Go2CruisingServicePy 创建')

        # 创建服务端
        self.service = self.create_service(Cruising, 'cruising', self.cru_cb)
        
        # 订阅里程计
        self.current_point = Point()
        self.odom_sub = self.create_subscription(Odometry, 'odom', self.odom_cb, 10)
         
        # 创建速度控制的发布对象
        self.api_id = ROBOT_SPORT_API_IDS['BALANCESTAND']
        self.req_pub = self.create_publisher(Request, '/api/sport/request', 10)
        self.timer = self.create_timer(0.1, self.on_timer) 

        # 本地参数
        self.declare_parameter('x', 0.1)
        self.declare_parameter('y', 0.0)
        self.declare_parameter('z', 0.5)

    # 服务回调函数
    def cru_cb(self, request: Cruising.Request, response: Cruising.Response):
        flag = request.flag

        # 判断巡航状态
        if flag != 0:
            api_id = ROBOT_SPORT_API_IDS['MOVE']
            self.get_logger().info('开始巡航...')
        else:
            api_id = ROBOT_SPORT_API_IDS['STOPMOVE']
            self.get_logger().info('停止巡航...')

        # 响应机器人的当前位置信息
        response.point = self.current_point
        return response

    # 里程计回调
    def odom_cb(self, odom: Odometry):
        self.current_point = odom.pose.pose.position

    def on_timer(self):
        req = Request()
        # 设置数据
        req.header.identity.api_id = self.api_id
        # 设置参数
        js = {
            # "x": self.get_parameter('x').get_parameter_value().double_value,
            "x": self.get_parameter('x').value,     # 推荐
            "y": self.get_parameter('y').value,
            "z": self.get_parameter('z').value
        }
        req.parameter = json.dumps(js)
        self.req_pub.publish(req)
    
def main(args=None):
    rclpy.init(args=args)

    node = Go2CruisingServicePy('go2_crusing_service_py')
    rclpy.spin(node)   
    rclpy.shutdown()


if __name__ == '__main__':
    main()
